src/dune_rules/merlin/ocaml_index.ml

```diff
@@ -0,0 +1,119 @@
+open Import
+module CC = Compilation_context
+module SC = Super_context
+
+let ocaml_index sctx ~dir =
+  Super_context.resolve_program_memo ~loc:None ~dir sctx "ocaml-index"
+;;
+
+let index_path_in_obj_dir ?for_cmt obj_dir =
+  let dir = Obj_dir.obj_dir obj_dir in
+  match for_cmt with
+  | Some path ->
+    let name = Path.basename path in
+    Path.Build.relative dir @@ Printf.sprintf ".index/%s.ocaml-index" name
+  | None -> Path.Build.relative dir "cctx.ocaml-index"
+;;
+
+let project_index ~build_dir = Path.Build.relative build_dir "project.ocaml-index"
+
+let cctx_rules cctx =
+  let open Memo.O in
+  (* Indexing is performed by the external binary [ocaml-index] which performs
+     full shape reduction to compute the actual definition of all the elements in
+     the typedtree. This step is therefore dependent on all the cmts of those
+     definitions are used by all the cmts of modules in this cctx. *)
+  let dir = CC.dir cctx in
+  let modules =
+    CC.modules cctx |> Modules.fold_no_vlib ~init:[] ~f:(fun x acc -> x :: acc)
+  in
+  let sctx = CC.super_context cctx in
+  let obj_dir = CC.obj_dir cctx in
+  let cm_kind = Lib_mode.Cm_kind.(Ocaml Cmi) in
+  let modules_with_cmts =
+    List.filter_map
+      ~f:(fun module_ ->
+        Obj_dir.Module.cmt_file obj_dir ~ml_kind:Impl ~cm_kind module_
+        |> Option.map ~f:(fun cmt -> Path.build cmt))
+      modules
+  in
+  let modules_with_cmtis =
+    List.filter_map
+      ~f:(fun module_ ->
+        Obj_dir.Module.cmt_file obj_dir ~ml_kind:Intf ~cm_kind module_
+        |> Option.map ~f:(fun cmti -> Path.build cmti))
+      modules
+  in
+  let* ocaml_index = ocaml_index sctx ~dir in
+  let context_dir =
+    CC.context cctx |> Context.name |> Context_name.build_dir |> Path.build
+  in
+  let* additionnal_libs =
+    (* The indexer relies on the load_path of cmt files. When
+       [implicit_transitive_deps] is set to [false] some necessary paths will
+       be missing.These are passed to the indexer with the `-I` flag.
+
+       The implicit transitive libs correspond to the set:
+       (requires_link \ requires_link)
+    *)
+    let open Resolve.Memo.O in
+    let* req_link = CC.requires_link cctx in
+    let+ req_compile = CC.requires_compile cctx in
+    let a =
+      List.fold_left req_link ~init:[] ~f:(fun acc l ->
+        if List.exists req_compile ~f:(Lib.equal l)
+        then acc
+        else (
+          let dir = Lib.info l |> Lib_info.obj_dir |> Obj_dir.byte_dir in
+          Command.Args.(A "-I" :: Path dir :: acc)))
+    in
+    Command.Args.S a
+  in
+  let fn = index_path_in_obj_dir obj_dir in
+  let includes =
+    Resolve.peek additionnal_libs |> Result.value ~default:Command.Args.empty
+  in
+  let aggregate =
+    Command.run
```

Comment:
This should be `Command.run_dyn_prog` and the resolution of the binary should be delayed into action builder